### ISSUE TITLE
Error when deserializing internally-tagged enum

### ISSUE THREAD

**Issue by alex-shapiro** [3Hren/msgpack-rust#130]
Given an internally-tagged enum with NewType variants, serializing produces a valid value but deserializing fails. For example:

```rust
#[derive(Serialize, Deserialize)]
#[serde(tag = "type")]
enum TestEnum {
    Foo(TestType),
    Bar(TestType),
}

#[derive(Serialize, Deserialize)]
struct TestType {
  inner: u32,
}

let test_enum1 = TestEnum::Foo(TestType{inner: 123});
let serialized = rmp_serde::to_vec(&test_enum1).unwrap();
let test_enum2 = rmp_serde::from_slice(&serialized).unwrap();
```
The serializer produces `[146, 163, 70, 111, 111, 123]`, equivalent to the JSON value `["Foo", 123]`. That's the value I was expecting. 

The deserializer fails with the error `Syntax("invalid type: sequence, expected any value")`.

Everything works if I remove the line `#[serde(tag = "type")]`, but then the serializer produces `[146, 0, 145, 145, 123]`, equivalent to the JSON value `[0, [[123]]]`. If possible I'd prefer to use the internally tagged representation.

**Comment by 3Hren** [comment]
I have no strong opinion of how enums should be serialized, but I think it's connected with #129 issue. Unfortunately currently I have no time to fix this (and possibly will have no time the next 2-3 weeks), however PR's are welcome!

**Comment by m4b** [comment]
I believe we have a similar issue, which caused us to switch :/ Unfortunately we simply don't have time to produce a repro, but this looks like it's the same as our issue (reading after writing fails, which imho should never ever occur for the same serializer/deserializer, no?)

Ref: https://github.com/das-labor/panopticon/pull/288#issuecomment-306343034

**Comment by SX91** [comment]
Internally tagged enums generates some strange code for `Enum::deserializer`:

```rust
let __tagged =
    match _serde::Deserializer::deserialize_any(
        __deserializer,
        _serde::private::de::TaggedContentVisitor::<__Field>::new("type"),
    ) {
        ::result::Result::Ok(val) => val,
        ::result::Result::Err(err) => {
            return ::result::Result::Err(::convert::From::from(err))
        }
    };
match __tagged.tag {
    __Field::__field0 => {
        _serde::export::Result::map(
            <Struct as _serde::Deserialize>::deserialize(
                _serde::private::de::ContentDeserializer::<__D::Error>::new(
                    __tagged.content,
                ),
            ),
            Enum::Foo,
        )
    }
    __Field::__field1 => {
        **match _serde::Deserializer::deserialize_any(**
            _serde::private::de::ContentDeserializer::<__D::Error>::new(
                __tagged.content,
            ),
            _serde::private::de::InternallyTaggedUnitVisitor::new(
                "Enum",
                "Bar",
            ),
        ) {
            ::result::Result::Ok(val) => val,
            ::result::Result::Err(err) => {
                return ::result::Result::Err(::convert::From::from(err))
            }
        };
        _serde::export::Ok(Enum::Bar)
    }
}
```

So actually `deserialize` tries to deserialize tag outside of struct (which is `sequence` in rmp) which results in the described error code.

**Comment by jD91mZM2** [comment]
I desperately need this... Any news?

> My guess: it's a `serde` bug

[Nope.](https://play.rust-lang.org/?gist=3c0891feff3526480b78943dd6dda8da&version=stable)

**Comment by SX91** [comment]
@jD91mZM2 what is this "nope" means? Did you try to investigate a little?

Look [here](https://github.com/serde-rs/serde/blob/25ab84d4b9df7ac0c7df361c9e543975cbebc502/serde/src/private/de.rs#L840). This is the code that actually parses the tagged structs. As you can see, only mappings are supported, while msgpack serializes structs as sequences. That's the whole reason for `Syntax("invalid type: sequence, expected any value")` error.
If you want support for representation described by @alex-shapiro (`["Foo", 123]`) to be parsed correctly, contact serde team for sequence support at `TaggedContentVisitor`.

**Comment by jD91mZM2** [comment]
> what is this "nope" means? Did you try to investigate a little?

It's a link. https://play.rust-lang.org/?gist=3c0891feff3526480b78943dd6dda8da&version=stable  
The error does not appear when using serde_json.

**Comment by SX91** [comment]
@jD91mZM2 well, read my comment then! It works with JSON because `serde_json` represents structures as mappings while `rmp_serde` represents them as sequences which is _unsupported_ by `serde` itself (`serde` supports _mapping_ only for internally tagged enums).

**Comment by SX91** [comment]
@3Hren I see only two ways to fix this: either `serde` extends support for internally tagged enums on sequences or Serializer/Deserializer could be modified to optionally allow representation of structs as mappings (binary JSON).

**Comment by jD91mZM2** [comment]
@SX91 Sorry, I didn't understand at first. Thank you for explaining :)

**Comment by 3Hren** [comment]
I'll try to do some template magic to make serializer/deserializer configurable.

**Comment by 3Hren** [comment]
For serializer I ended up with recursive types (Serializer depends from SerializeStruct, which depends from Serializer), which obviously won't compile. Using trait objects resolves this issue, but I this hurts by disalowing inlining and adds one indirect call.
Have to try again...

**Comment by 3Hren** [comment]
Ok, for extending serialization I ended up with two different solutions:
- One serializer, but customizable with template parameters (https://github.com/3Hren/msgpack-rust/commit/5f8568c4f3eb885d4d3fdeac2e597b0cb8ac2de8#diff-d6caf16ec62e16ed5351ec4bd0a1cc58R88).
  - Simplier, less code.
- Many serializer wrappers (https://github.com/3Hren/msgpack-rust/commit/14ebba4820d85c637eb09d8448e166b541672ce3).
  - Do not break ABI on extensions.

Personally I prefer the second solution, but currently rust can't compile such code, trying to instantiate infinite type for no reasons.

**Comment by 3Hren** [comment]
Hmm, it works! See https://github.com/3Hren/msgpack-rust/pull/146 for progress.

**Comment by 3Hren** [comment]
And again I'm in trouble.

@dtolnay how can I properly implement `Serializer` wrappers? Imagine we have some core `CoreSerializer` and we want to override exactly one method (for example `serialize_struct`). We create `WrapSerializer<S>` where `S` - another serializer and delegate all methods to him, except `serialize_struct`. Also we want to have access to the underlying `W: Write` object to be able to write directly.

I tried and failed, because of many reasons: weird compilator errors with infinite type instantiations (actually some conditions were not met), lifetimes issues, requirements to have both `&'a mut Serializer` and `Serializer` implementations for wrapper, etc.

**Comment by dtolnay** [comment]
I wish someone had reported this to the Serde repo earlier! I released a fix in serde_derive 1.0.15. The test code in https://github.com/3Hren/msgpack-rust/issues/130#issuecomment-316758123 now works.

**Comment by 3Hren** [comment]
Many thanks! But my question about serializer wrappers is still actual :)

**Comment by dtolnay** [comment]
> We create `WrapSerializer<S>` where `S` - another serializer and delegate all methods to him, except `serialize_struct`. Also we want to have access to the underlying `W: Write` object to be able to write directly.

I don't think this works in general. Not all serializers have an underlying Write object. Not all serializers even produce bytes as output. [Here is one counterexample.](https://github.com/serde-rs/json/blob/v1.0.3/src/value/ser.rs#L41-L45)

You can either wrap an arbitrary Serializer type and interact with it through only the Serializer interface ([example](https://github.com/dtolnay/erased-serde/blob/0.3.1/src/ser.rs#L222-L224)), or wrap a concrete Serializer and access whatever APIs are visible to your code ([example](https://github.com/serde-rs/json/blob/v1.0.3/src/ser.rs#L831-L833)).

**Comment by 3Hren** [comment]
> interact with it through only the Serializer interface

Yep, and access to the writer by requiring the `S` of `WrapperSerializer<S>` to impl some kind of `GetWrite` trait. But still no success...
Well, I'll continue trying, but maybe someone smarter than me succeeds.

**Comment by dtolnay** [comment]
Can you share the nonworking code?

**Comment by 3Hren** [comment]
The idea is:

```rust
pub struct CoreSerializer<W> {
    wr: W,
}

pub trait UnderlyingWrite {
    type Write: Write;

    fn get_mut(&self) -> &mut Self::Write;
}

impl<W: Write> UnderlyingWrite for Serializer<W> {
    type Write = W;

    fn get_mut(&mut self) -> &mut W {
        &mut self.wr
    }
}

pub struct WrapperSerializer<S> {
    se: S,
}

impl<'a, S, W> UnderlyingWrite for &'a mut WrapperSerializer<S>
where
    S: UnderlyingWrite<Write = W>,
    W: Write
{
    type Write = W;

    fn get_mut(&mut self) -> &mut S::Write {
        self.se.get_mut()
    }
}

impl<'a, S> Serializer for &'a mut WrapperSerializer<S>
where
    S: UnderlyingWrite,
    &'a mut S: Serializer<Ok = (), Error = Error>
{
    type Ok = ();
    type Error = Error;

    type SerializeSeq = <&'a mut S as Serializer>::SerializeSeq;
    type SerializeTuple = <&'a mut S as Serializer>::SerializeTuple;
    type SerializeTupleStruct = <&'a mut S as Serializer>::SerializeTupleStruct;
    type SerializeTupleVariant = <&'a mut S as Serializer>::SerializeTupleVariant;
    type SerializeMap = <&'a mut S as Serializer>::SerializeMap;
    type SerializeStruct = Self;
    type SerializeStructVariant = <&'a mut S as Serializer>::SerializeStructVariant;

    // Delegate all methods except `serialize_struct` to self.se.

    #[inline]
    fn serialize_struct(self, name: &'static str, len: usize) -> Result<Self::SerializeStruct, Self::Error> {
        // Do our magic.
        Ok(self)
    }
}

impl<'a, S> SerializeStruct for &'a mut WrapperSerializer<S>
where
    S: UnderlyingWrite,
    &'a mut S: Serializer<Ok = (), Error = Error>
{
    type Ok = ();
    type Error = Error;

    fn serialize_field<T: ?Sized + Serialize>(&mut self, key: &'static str, value: &T) ->
        Result<(), Self::Error>
    {
        // Do more our magic, for example rmp::encode::write_str(self.se.get_mut(), key)?;
        value.serialize(&mut **self) // Here it bangs.
    }

    fn end(self) -> Result<Self::Ok, Self::Error> {
        Ok(())
    }
}
```

**Comment by 3Hren** [comment]
HKTB helped to compile the part of code at least:

```rust
impl<'a, S> Serializer for &'a mut StructMapSerializer<S>
where
    S: UnderlyingWrite,
    for<'b> &'b mut S: Serializer<Ok = (), Error = Error>
{
...
}
```

Invertigating more...

**Comment by dtolnay** [comment]
Here are the types in `serialize_field`:

- The type of `Self` is `&'a mut WrapperSerializer<S>`.
- The type of `self` is `&mut Self` which is `&mut &'a mut WrapperSerializer<S>`.
- In order for the argument to value::serialize to implement Serializer, the argument must be `&'a mut WrapperSerializer<S>`.

The reason it fails [has to do with variance of &mut references](https://doc.rust-lang.org/nomicon/subtyping.html).

```rust
// Okay.
fn f(s: &mut &'static str) {
    let _: &'static str = &**s;
}

// Not okay.
fn g(s: &mut &'static mut str) {
    let _: &'static mut str = &mut **s;
}
```

**Comment by 3Hren** [comment]
Thanks for the explanation!

After ~2 years of Rust programming backed with 6 years of C++ template programming I still have no idea what I'm doing.

Using HKTB (not sure it's the proper solution) my idea works, everything is green now.

**Comment by alex-shapiro** [comment]
I'm closing this issue because it appears to be fixed by #146

**Comment by jD91mZM2** [comment]
I don't want to nitpick, but I noticed it serializes it using the name now instead of the normal id. This isn't worse by any means, but it's... different. Any chance it could somehow be configured?
